### ISSUE TITLE
CrossEntropyLoss, BCELoss and BCEWithLogitsLoss use the unstable compositions PyTorch avoids: inf loss (and NaN grads) where the reference returns a finite value

### ISSUE THREAD

**Issue by botify-service** [tenstorrent/tt-forge-fe#3405]
### Summary

Three loss functions in `forge/forge/op/loss.py` are built from the exact compositions that
PyTorch's own implementations avoid for numerical stability. All three return `inf` — and
therefore `NaN` gradients — on inputs where the reference returns an ordinary finite loss.

Each divergence below is checked against PyTorch's implementation in
`aten/src/ATen/native/Loss.cpp` rather than against a tolerance, so none of them is an accuracy
argument.

---

### 1. `CrossEntropyLoss` forms the probability before taking the log

```python
# loss.py, CrossEntropyLoss.forward
softmax = Softmax("softmax", predictions, dim=-1)
log_softmax = Log("log", softmax)
product = Multiply("products", labels, log_softmax)
```

`softmax` underflows to exactly `0` for any logit far enough below the row max, and `log(0)` is
`-inf`. PyTorch computes `log_softmax` as `x - logsumexp(x)` and never materialises the
probability.

`float32`, one-hot label on the smallest logit:

| logits | forge loss | stable loss |
|---|---|---|
| `[10, 0, -100]` | **`inf`** | `110.000045` |
| `[50, 0, -120]` | **`inf`** | `170.000000` |
| `[2, 1, 0.5]` | `1.96437` | `1.964369` — control, matches |

`110` and `170` are ordinary `float32` values. The control row shares the same code path and
passes, so this is not general breakage.

The threshold is the `exp` underflow point — a logit gap of roughly 88 in `float32`, less in
`bfloat16`. A trained classifier over a large vocabulary opens that gap routinely, so this
appears partway through training rather than at step 0.

---

### 2. `BCELoss` omits the clamp that PyTorch documents, and uses the less accurate `log(1-p)`

```python
# loss.py, BCELoss.forward
log_prediction = Log("log", prediction)                       # log(p)
one_minus_prediction = Subtract("one_minus_prediction", one, prediction)
log_one_minus_prediction = Log("log_one_minus_prediction", one_minus_prediction)   # log(1-p)
```

PyTorch's kernel, for the same equation:

```cpp
// aten/src/ATen/native/Loss.cpp
// Binary cross entropy tensor is defined by the equation:
// L = -w (y ln(x) + (1-y) ln(1-x))
return (target_val - scalar_t(1)) *
    std::max(scalar_t(std::log1p(-input_val)), scalar_t(-100)) -
    target_val *
    std::max(scalar_t(std::log(input_val)), scalar_t(-100));
```

Two differences, both deliberate on PyTorch's side:

- **both logs are clamped to `>= -100`**, which is what guarantees a finite loss at `p = 0` or
  `p = 1`;
- **`log1p(-p)` is used instead of `log(1 - p)`**, which keeps precision when `p` is small and
  `1 - p` rounds to `1`.

| p | y | forge | PyTorch (clamped) |
|---|---|---|---|
| `0.0` | 1 | **`inf`** | `100.0000` |
| `1.0` | 0 | **`inf`** | `100.0000` |
| `1e-45` | 1 | `103.616` | `100.0000` |
| `0.5` | 1 | `0.693147` | `0.6931` — control, matches |

`p = 0` and `p = 1` are not exotic: a saturated sigmoid produces them in `bfloat16` well before
the underlying logit is extreme.

---

### 3. `BCEWithLogitsLoss` is the composition it exists to replace

```python
class BCEWithLogitsLoss(ForgeModule):
    """loss = BCELoss(Sigmoid(predictions), labels)"""
    sigmoid_prediction = Sigmoid("sigmoid", prediction)
    loss = self.bce_loss(sigmoid_prediction, labels)
```

PyTorch's version never forms the sigmoid:

```cpp
Tensor binary_cross_entropy_with_logits(...) {
  auto log_sigmoid_input = at::log_sigmoid(input);
  ...
  Tensor loss = (1 - target).mul_(input).sub_(log_sigmoid_input);
```

That is `(1 - y) * x - log_sigmoid(x)`, the log-sum-exp form. The whole reason
`BCEWithLogitsLoss` is a separate class from `Sigmoid` + `BCELoss` is to avoid the round trip
through the probability — and this implementation performs exactly that round trip, then feeds
the result into the unclamped `BCELoss` above, so it inherits defect 2 as well.

| logit | y | sigmoid(x) | forge | stable |
|---|---|---|---|---|
| `-100` | 1 | `0` | **`inf`** | `100.000000` |
| `100` | 0 | `1` | **`inf`** | `100.000000` |
| `-50` | 1 | `1.93e-22` | `49.99999998` | `50.000000` |
| `2` | 1 | `0.880797` | `0.126928` | `0.126928` — control |

`sigmoid` saturates to exactly `0` or `1` around `|x| ~ 89` in `float32` and far earlier in
`bfloat16`, so the `inf` is reachable from ordinary logits.

---

### Expected

- `CrossEntropyLoss` should use a stable `log_softmax` — `x - logsumexp(x, dim=-1)` — rather
  than `Log(Softmax(x))`.
- `BCELoss` should clamp both logs at `-100` to match the reference, and should use a `log1p`
  form for `log(1 - p)` if one is available.
- `BCEWithLogitsLoss` should be built on `log_sigmoid` directly — `(1 - y) * x - log_sigmoid(x)`
  — instead of delegating to `Sigmoid` + `BCELoss`.

`log_sigmoid` already exists in the stack, so all three have the primitives available.

### Minimum reproducible cases

```python
import forge, torch

# 1. CrossEntropyLoss
predictions = torch.tensor([[10.0, 0.0, -100.0]])
labels      = torch.tensor([[0.0, 0.0, 1.0]])       # one-hot on the smallest logit
# expected ~110.000045 ; today: inf

# 2. BCELoss
prediction = torch.tensor([[0.0]])
labels     = torch.tensor([[1.0]])
# expected 100.0 (PyTorch clamps at -100) ; today: inf

# 3. BCEWithLogitsLoss
prediction = torch.tensor([[-100.0]])
labels     = torch.tensor([[1.0]])
# expected 100.0 ; today: inf
```

Each has a control that must keep passing in the same run: `[[2.0, 1.0, 0.5]]` for (1),
`p = 0.5` for (2), and `logit = 2.0` for (3).

### Why the current tests do not catch it

The loss tests use small, well-conditioned inputs — logits in a narrow band and probabilities
away from `0` and `1` — where all three compositions agree with the stable forms to within
tolerance. Nothing samples a saturated probability or a large logit gap, which is the only
condition that separates them.

### Impact

These are loss functions, so the failure does not stay local: an `inf` loss produces `NaN`
gradients and the optimizer state is corrupted from that step onward. The trigger is a
*confident* model, which means training can run normally for a long stretch and then diverge with
no obvious cause — the hardest failure shape to attribute.

### Notes on verification

Source analysis plus exact arithmetic; not run on hardware. The forge compositions are readable at
the lines quoted, and the PyTorch behaviour is quoted from `aten/src/ATen/native/Loss.cpp` rather
than from documentation. The tables were produced by evaluating both formulations in `float32`
with each intermediate rounded and flush-to-zero modelled on the underflowing `exp`/`sigmoid`; the
`inf` boundaries follow from `float32` semantics alone, so they should reproduce on device.
